ImageData/Hdf5Loader.h

```diff
@@ -293,13 +296,138 @@ bool Hdf5Loader::GetCursorSpectralData(std::vector<float>& data, int stokes, int
+        return nullptr;
+    }
+
+    int num_z = _num_channels;
+
+    bool recalculate(false);
+
+    if (_region_stats.find(region_id) == _region_stats.end()) { // region stats never calculated
+        _region_stats.emplace(
+            std::piecewise_construct, std::forward_as_tuple(region_id), std::forward_as_tuple(origin, mask->shape(), num_z));
+        recalculate = true;
+    } else if (!_region_stats[region_id].IsValid(origin, mask->shape())) { // region stats expired
+        recalculate = true;
+    }
+
+    if (recalculate) {
+        int num_y = mask->shape()(0);
+        int num_x = mask->shape()(1);
+
+        int x_min = origin(0);
+        int x_max = x_min + num_x;
+        int y_min = origin(1);
+        int y_max = y_min + num_y;
+
+        auto& stats = _region_stats[region_id].stats;
+
+        auto& num_pixels = stats[CARTA::StatsType::NumPixels];
+        auto& nan_count = stats[CARTA::StatsType::NanCount];
+        auto& sum = stats[CARTA::StatsType::Sum];
+        auto& mean = stats[CARTA::StatsType::Mean];
+        auto& rms = stats[CARTA::StatsType::RMS];
+        auto& sigma = stats[CARTA::StatsType::Sigma];
+        auto& sum_sq = stats[CARTA::StatsType::SumSq];
+        auto& min = stats[CARTA::StatsType::Min];
+        auto& max = stats[CARTA::StatsType::Max];
+
+        std::vector<float> slice_data(num_z * num_y);
+
+        // Set initial values of stats which will be incremented (we may have expired region data)
+        for (size_t z = 0; z < num_z; z++) {
+            min[z] = FLT_MAX;
+            max[z] = FLT_MIN;
+            num_pixels[z] = 0;
+            nan_count[z] = 0;
+            sum[z] = 0;
+            sum_sq[z] = 0;
+        }
+
+        // Load each X slice of the swizzled region bounding box and update Z stats incrementally
+        for (size_t x = 0; x < num_x; x++) {
+            casacore::Slicer slicer;
+            if (_num_dims == 4) {
+                slicer = casacore::Slicer(IPos(4, 0, y_min, x + x_min, stokes), IPos(4, num_z, num_y, 1, 1));
+            } else if (_num_dims == 3) {
+                slicer = casacore::Slicer(IPos(3, 0, y_min, x + x_min), IPos(3, num_z, num_y, 1));
+            }
+
+            casacore::Array<float> tmp(slicer.length(), slice_data.data(), casacore::StorageInitPolicy::SHARE);
+
+            try {
+                LoadSwizzledData(FileInfo::Data::SWIZZLED)->doGetSlice(tmp, slicer);
+            } catch (casacore::AipsError& err) {
+                std::cerr << "Could not load cursor spectral data from swizzled HDF5 dataset. AIPS ERROR: " << err.getMesg() << std::endl;
+                return nullptr;
+            }
+
+            for (size_t y = 0; y < num_y; y++) {
+                // skip all Z values for masked pixels
+                if (!mask->getAt(IPos(2, x, y))) {
+                    continue;
+                }
+                for (size_t z = 0; z < num_z; z++) {
+                    float& v = slice_data[y * num_z + z];
```

Comment:
no need to use a reference for a primitive type. In fact, since all the stats are done using doubles, it would be better to just use `double v`